python_modules/libraries/dagster-components/dagster_components/core/schema/base.py

```diff
@@ -104,52 +96,24 @@ def from_annotation(annotation: Any, field_name: str) -> "FieldResolver":
         )
 
 
-@record
-class FieldResolverInfo:
-    field_name: str
-    resolver: FieldResolver
-
-
-def field_resolver(field_name: str) -> Any:
-    def decorator(
-        fn: Callable[["ResolutionContext", ResolvableSchema], T],
-    ) -> Callable[["ResolutionContext", ResolvableSchema], T]:
-        setattr(
-            _get_annotation_target(fn),
-            FIELD_RESOLVER_ATTR,
-            FieldResolverInfo(field_name=field_name, resolver=FieldResolver(fn=fn)),
-        )
-        return fn
-
-    return decorator
-
-
-def _get_field_resolver_info(obj: Any) -> Optional[FieldResolverInfo]:
-    return getattr(_get_annotation_target(obj), FIELD_RESOLVER_ATTR, None)
-
-
 def _get_annotation_field_resolvers(cls: type) -> dict[str, FieldResolver]:
     if issubclass(cls, BaseModel):
-        field_names = cls.model_fields.keys()
+        # neither pydantic's Field.annotation nor Field.rebuild_annotation() actually rerturn
+        # the original annotation, so we have to walk the mro to get them
+        annotations = {}
+        for field_name in cls.model_fields:
+            for base in cls.__mro__:
+                if field_name in getattr(base, "__annotations__", {}):
+                    annotations[field_name] = base.__annotations__[field_name]
+                    break
+        return {
+            field_name: FieldResolver.from_annotation(annotation, field_name)
+            for field_name, annotation in annotations.items()
+        }
     elif is_dataclass(cls):
-        field_names = {field.name for field in fields(cls)}
+        return {
+            field.name: FieldResolver.from_annotation(field.type, field.name)
+            for field in fields(cls)
+        }
```

Comment:
sort of tangential: what happens when we fall through to this else case?

Reply:
In the future, we could have a third branch that inspects the `__init__` args of the type (if it has an init fn), but for now we fall back on grabbing annotations from the ResolvedSchema class itself (with the assumption that its field names align with the target class field names)